stac_model/utils.py

```
@@ -0,0 +1,215 @@
+from typing import Optional, cast
+
+import torch.nn as nn
+from pystac import Asset, Item, Link
+from pystac.extensions.eo import Band, EOExtension
+
+from stac_model.base import TaskEnum
+from stac_model.input import InputStructure, ModelInput
+from stac_model.output import MLMClassification, ModelOutput, ModelResult
+from stac_model.schema import ItemMLModelExtension, MLModelExtension, MLModelProperties
+
+
+def get_input_channels(state_dict: dict) -> int:
+    """
+    Get number of input channels from the first convolutional layer's weights.
+    """
+    for key, tensor in state_dict.items():
+        if "encoder._conv_stem.weight" in key:
+            return tensor.shape[1]
+    raise ValueError("Could not determine input channels from model weights.")
+
+
+def get_output_channels(state_dict: dict) -> int:
+    """
+    Get number of output channels from the segmentation head's last conv layer.
+    """
+    for key, tensor in reversed(state_dict.items()):
+        if "segmentation_head.0.weight" in key:
+            return tensor.shape[0]
+    raise ValueError("Could not determine output channels from model weights.")
+
+
+def from_torch(
+    model: nn.Module,
+    *,
+    weights: Optional[object] = None,
+    item_id: str = "torch-model",
+    bbox: Optional[list[float]] = None,
+    geometry: Optional[dict] = None,
+    links: Optional[list[dict]] = None,
+    datetime_range: tuple[str, str] = (
+        "2015-06-23T00:00:00Z",  # Sentinel-2A launch date (first Sentinel-2 data available)
+        "2024-08-27T23:59:59Z",  # Dataset publication date Fields of The World (FTW)
```

These example defaults should not be used here. The field should have `None` default and raise if they are not provided.